File: posawesome/api/logger.py
# -*- coding: utf-8 -*-
"""
Frontend Logger API
Receives logs from frontend and writes them to posawesome_info.log and posawesome_error.log
"""
from __future__ import unicode_literals
import frappe
from posawesome import info_logger, error_logger


@frappe.whitelist()
def log_frontend(level, file, message, data=None):
    """
    Log frontend messages to backend log files

    Args:
        level: Log level (info, warn, error)
        file: Frontend file name
        message: Log message
        data: Additional data (JSON string)
    """
    try:
        log_message = f"[{file}] {message}"

        if data:
            log_message += f" | Data: {data}"

        # Log based on level
        if level == 'error':
            error_logger.error(log_message)
        elif level == 'warn':
            error_logger.warning(log_message)
        else:
            info_logger.info(log_message)

        return {"success": True}

    except Exception as e:
        frappe.log_error(
            title=f"Frontend Logger Error",
            message=str(e)
        )
        return {"success": False, "error": str(e)}

File: posawesome/public/js/posapp/logger.js
/**
 * POSAwesome Frontend Logger
 * Logs messages to both console and backend via frappe.call
 */

export const posawesome_logger = {
  /**
   * Log info message
   * @param {string} file - Filename where log is triggered
   * @param {string} message - Log message
   * @param {object} data - Additional data to log
   */
  info(file, message, data = null) {
    const logMessage = `[${file}] ${message}`;
    console.log(logMessage, data || "");

    // Send to backend for persistent logging
    this._sendToBackend("info", file, message, data);
  },

  /**
   * Log error message
   * @param {string} file - Filename where error occurred
   * @param {string} message - Error message
   * @param {object} error - Error object or additional data
   */
  error(file, message, error = null) {
    const logMessage = `[${file}] ${message}`;
    console.error(logMessage, error || "");

    // Send to backend for persistent logging
    this._sendToBackend("error", file, message, error);
  },

  /**
   * Log warning message
   * @param {string} file - Filename where warning occurred
   * @param {string} message - Warning message
   * @param {object} data - Additional data to log
   */
  warn(file, message, data = null) {
    const logMessage = `[${file}] ${message}`;
    console.warn(logMessage, data || "");

    // Send to backend for persistent logging
    this._sendToBackend("warn", file, message, data);
  },

  /**
   * Send log to backend
   * @private
   */
  _sendToBackend(level, file, message, data) {
    // Only send errors and important warnings to backend
    if (level === "error" || (level === "warn" && data)) {
      frappe.call({
        method: "posawesome.api.logger.log_frontend",
        args: {
          level: level,
          file: file,
          message: message,
          data: data ? JSON.stringify(data) : null,
        },
        freeze: false,
        async: true,
      });
    }
  },
};

// Make it globally available
window.posawesome_logger = posawesome_logger;
